### ISSUE TITLE
Treat custom elements and native DOM elements uniformly, decouple binding interface from binding implementation

### ISSUE THREAD

**Issue by futpib** [knockout/knockout#2169]
Custom elements are just components, and knockout makes clear distinction between components and native DOM elements.

This makes it difficult to use a custom element as a drop-in replacement for a native element. Especially when a custom binding provider is involved, but even without it.

There is also an issue with binding implementation. It has to handle all possible elements it might be bound to.

This bloats bindings code with special checks for different kinds of elements. And sometimes bloats other bindings code with checks like `if (allBindings.has('value')) { /* something special */ }` that feels almost like monkey-patching the `value` (or other) binding.

Let me demonstrate:

Say we have a code that uses `select` with some pretty standard bindings:
```html
<select data-bind="
    options: options,
    value: value,
    disable: disable,
    visible: visible
"></select>
```
Now say we want to use a custom `bootstrap-select` instead, let's imagine we did a stub implementation of a `bootstrap-select` component and simply changed the tag name for starters:
```html
<bootstrap-select data-bind="
    options: options,
    value: value,
    disable: disable,
    visible: visible
"></bootstrap-select>
```

The only binding that would work properly is `visible`, this is one of the few built-in bindings that are accidentally generic, it obviously would work because every element can be hidden with `style="display: none;"`. Few others are `attr`, `event` and `css`.

To complete our implementation we would have to implement `options`, `value` and `disable` parameters in our component.
```html
<bootstrap-select
    params="
        options: options,
        value: value,
        disable: disable
    "
    data-bind="
        visible: visible
    "
></bootstrap-select>
```

**tl;dr**:
I feel like it would have been easier if the following were true:
1. Binding name, arguments (signature?) and meaning were separated from binding implementation
2. There was no distinction between component's parameters and bindings (`params` and `data-bind`)
3. Appropriate implementation was used for each native or custom element.

Obviously leaving it to the user to implement whichever bindings are supported by a given custom element. 
Optionally. if a binding is not implemented for a given element an error is thrown (as per  #1786 ).

This would also result in a cleaner implementation of bindings like `value` (it currently handles every type of `input` as well as `select`)

This will also mitigate occasional need for checks like
```js
// in a binding handler
if (allBindings.has('value')) {
    // bind custom `change`-like events
    // write code that updates element from model using custom value setter method
}
```
This happens almost every time we have to write a binding for a jQuery plugin (typeahead, datepicker, inputmask, etc.).

**Comment by mbest** [comment]
Related: #1444
